<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="tower_yaw">


<!-- GEOMETRY -->
    <xacro:macro name="tower_yaw_geometry">
        <geometry>
            <cylinder length="${tower_yaw_height}" radius="${tower_yaw_radius}"/>
        </geometry>
    </xacro:macro>

<!-- MATERIAL -->
    <xacro:macro name="tower_yaw_material">
        <material name="dark_gray"/>
    </xacro:macro>

<!-- LINK -->
    <xacro:macro name="tower_yaw" params="name parent x y z">
        <link name="${name}">
            <visual>
                <xacro:tower_yaw_geometry/>
                <xacro:tower_yaw_material/>
            </visual>
            <collision>
                <xacro:tower_yaw_geometry/>
            </collision>
            <inertial>
                <mass value="1"/>
                <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.01"/>
            </inertial>
        </link>

            <joint name="${name}_joint" type="revolute">
                <parent link="${parent}"/>
                <child link="${name}"/>
                <origin rpy="0 0 0" xyz="${x} ${y} ${z}"/>
                <axis xyz="0 0 -1"/>
                <limit effort="100" velocity="10" lower="-3.14" upper="3.14"/>
                <dynamics damping="1"/>
            </joint>

            <gazebo reference="${name}">
                <mu1> .2 </mu1>
                <mu2> .2 </mu2>
                <material> Gazebo/Grey </material>
                <selfCollide> True </selfCollide>
            </gazebo>

            <gazebo>
                <plugin name="joint_limiter_${name}" filename="libJointLimiter.so">
                    <joint>${name}_joint</joint>
                    <topic>/simulation/controller/vel/${name}</topic>
                    <lower>-3.14</lower>
                    <upper>3.14</upper>
                </plugin>
            </gazebo>
            <gazebo>
                <plugin name="joint_controller_${name}" filename="libControllerPID.so">
                    <joint>${name}_joint</joint>
                    <topic>/simulation/controller/vel/${name}</topic>
                    <Kp>1</Kp>
                    <Ki>0.7</Ki>
                    <Kd>0.4</Kd>
                </plugin>
            </gazebo>

            <!--transmission name="${name}_transmission">
                <type>transmission_interface/SimpleTransmission</type>
                <joint name="${name}_joint">
                    <hardwareInterface>EffortJointInterface</hardwareInterface>
                </joint>
                <actuator name="${name}_motor">
                    <hardwareInterface>EffortJointInterface</hardwareInterface>
                    <mechanicalReduction>1</mechanicalReduction>
                </actuator>
            </transmission-->

    </xacro:macro>

</robot>
